contributors/design-proposals/node/limit-node-object-self-control.md

```diff
@@ -0,0 +1,127 @@
+has two problems that a centralized approach would not have. Minorly, it makes
+management difficult. Instead of configuring labels and taints in a centralized
+place, we must configure `N` kubelet command lines. More significantly, the
+approach greatly compromises security. Below are two straightforward escalations
+on an initially compromised node that exhibit the attack vector.
+
+### Capturing Dedicated Workloads
+
+Suppose company `foo` needs to run an application that deals with PII on
+dedicated nodes to comply with government regulation. A common mechanism for
+implementing dedicated nodes in Kubernetes today is to set a label or taint
+(e.g. `foo/dedicated=customer-info-app`) on the node and to select these
+dedicated nodes in the workload controller running `customer-info-app`.
+
+Since the nodes self reports labels upon registration, an intruder can easily
+register a compromised node with label `foo/dedicated=customer-info-app`. The
+scheduler will then bind `customer-info-app` to the compromised node potentially
+giving the intruder easy access to the PII.
+
+This attack also extends to secrets. Suppose company `foo` runs their outward
+facing nginx on dedicated nodes to reduce exposure to the company's publicly
+trusted server certificates. They use the secret mechanism to distribute the
+serving certificate key. An intruder captures the dedicated nginx workload in
+the same way and can now use the node certificate to read the company's serving
+certificate key.
+
+## Proposed Solution
+
+In many environments, we can improve the situation by centralizing reporting of
+sensitive node attributes to a more trusted source and disallowing reporting of
+these attributes from the kubelet.
+
+### Label And Taint Restriction
+
+An operator will configure a whitelist of taints and labels that nodes are
+allowed to set on themselves. This list should include the taints and labels
+that the kubelet is already setting on itself.
+
+Well known taint keys:
+```
+node.kubernetes.io/not-ready
+node.alpha.kubernetes.io/notReady
+node.kubernetes.io/unschedulable
+node.kubernetes.io/out-of-disk
+node.kubernetes.io/memory-pressure
+node.kubernetes.io/disk-pressure
+node.kubernetes.io/network-unavailable
+node.kubernetes.io/pid-pressure
+node.cloudprovider.kubernetes.io/uninitialized
+```
+
+Well known label keys:
+
+```
+kubernetes.io/hostname
+failure-domain.beta.kubernetes.io/zone
+failure-domain.beta.kubernetes.io/region
+beta.kubernetes.io/instance-type
+beta.kubernetes.io/os
+beta.kubernetes.io/arch
+```
+
+As well as any taints and labels that the operator is setting using:
+
+```
+  --register-with-taints
+  --node-labels
+```
+
+This whitelist is passed as a command line flag to the apiserver. Admission
+control (either NodeRestriction or in a new admission control) will then prevent
+setting and modification by nodes of all taints and labels with keys not in the
+whitelist.
```

Comment:
This means that if an administrator wants to add a new node pool with a new taint or label, they need to restart the APIserver. Is there an argument against dynamic config?

EDIT: I guess the work around is to just manually (or via an external component) label the nodes after they're registered?

Reply:
> I guess the work around is to just manually (or via an external component) label the nodes after they're registered?

I'd actually consider that the main recommendation, not a workaround

Reply:
maybe a configure file or confmap is better for dynamic update :)